src/test/java/org/jenkinsci/plugins/gitclient/GitUsernamePasswordBindTest.java

```diff
@@ -0,0 +1,72 @@
+package org.jenkinsci.plugins.gitclient;
+
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.CredentialsScope;
+import com.cloudbees.plugins.credentials.domains.Domain;
+import com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl;
+import hudson.FilePath;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+public class GitUsernamePasswordBindTest {
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[][]{
+                {"randomName", "special%%_342@**"}, {"here's-a-quote", "&Ampersand"}, {"semi;colon", "colon:inside"}
+        });
+    }
+
+    @Rule public final TemporaryFolder tempFolder = new TemporaryFolder();
+
+    @Rule public JenkinsRule r = new JenkinsRule();
+
+
+    private final String username;
+
+    private final String password;
+
+    public GitUsernamePasswordBindTest(String username, String password){
+        this.username = username;
+        this.password = password;
+    }
+
+    private File rootDir = null;
+    private FilePath rootFilePath = null;
+    private UsernamePasswordCredentialsImpl credentials=null;
+    private GitUsernamePasswordBind gitCredBind=null;
+
+    @Before
+    public void basicSetup() throws IOException {
+        //File init
+        rootDir = tempFolder.getRoot();
+        rootFilePath = new FilePath(rootDir.getAbsoluteFile());
+
+        //Credential init
+        credentials = new UsernamePasswordCredentialsImpl(CredentialsScope.GLOBAL, "80y7g2fjekru8h9237ywes2", "GIt Username and Password Binding Test", this.username, this.password);
+        CredentialsProvider.lookupStores(r.jenkins).iterator().next().addCredentials(Domain.global(), credentials);
+        gitCredBind = new GitUsernamePasswordBind(credentials.getId());
+    }
+
+    @Test
+    public void test_GenerateGitScript_write() throws IOException, InterruptedException {
+        GitUsernamePasswordBind.GenerateGitScript tempGenScript = new GitUsernamePasswordBind.GenerateGitScript(this.username,this.password,credentials.getId());
+        FilePath tempSriptFile = tempGenScript.write(credentials,rootFilePath);
+        assertNotNull(tempGenScript);
+        assertTrue("Username not found",tempSriptFile.readToString().contains(this.username));
+        assertTrue("Password not found",tempSriptFile.readToString().contains(this.password));
+    }
+}
```

Comment:
I was feeling like writing an automated test.  Here is a test that is largely copied from UsernamePasswordMultiBindingTest in the credentials binding plugin.  It needs more work because it relies on shell expansion of the environment variable to generate the test output.  The BatchFile and Shell builders could both be replaced with a call to an Ant builder, but I ran out of time to make that change.
```suggestion

    @Test
    public void test_FreeStyleProject() throws Exception {
        FreeStyleProject p = r.createFreeStyleProject();
        List<? extends MultiBinding<?>> bindings = Collections.<MultiBinding<?>>singletonList(new GitUsernamePasswordBind(credentials.getId()));
        SecretBuildWrapper secretWrapper = new SecretBuildWrapper(bindings);
        p.getBuildWrappersList().add(secretWrapper);
        if (isWindows()) {
            p.getBuildersList().add(new BatchFile("@echo off\necho %" + GitUsernamePasswordBind.USERNAMEKEY + "%:%" + GitUsernamePasswordBind.PASSWORDKEY + "% > auth.txt"));
        } else {
            p.getBuildersList().add(new Shell("set +x\necho \"$" + GitUsernamePasswordBind.USERNAMEKEY + ":$" + GitUsernamePasswordBind.PASSWORDKEY + "\" > auth.txt"));
        }
        r.configRoundtrip((Item) p);
        SecretBuildWrapper wrapper = p.getBuildWrappersList().get(SecretBuildWrapper.class);
        FreeStyleBuild b = r.buildAndAssertSuccess(p);
        r.assertLogNotContains(this.password, b);
        assertThat(b.getWorkspace().child("auth.txt").readToString().trim(), is(username + ':' + password));
        assertThat(new TreeSet<>(b.getSensitiveBuildVariables()).toString(), is("[" + GitUsernamePasswordBind.PASSWORDKEY + ", " + GitUsernamePasswordBind.USERNAMEKEY + "]"));
    }

    /**
     * inline ${@link hudson.Functions#isWindows()} to prevent a transient
     * remote classloader issue
     */
    private static boolean isWindows() {
        return File.pathSeparatorChar == ';';
    }

}
```

Reply:
Will work on replacing the BatchFile and Shell builders with an Ant builder and update further

Reply:
@arpoch I think that I've realized there is an easy way to make the test repeatable and reliable and still use the `Shell` and `Bat` steps.  No need to spend time on the Ant builder idea.  It would be very complicated, while the technique I'm envisioning feels simple.

The idea is to use shell steps like this:

```
new Shell("env | grep " + GitUsernamePasswordBind.PASSWORDKEY + "' > auth ; " +
                 "env | grep " + GitUsernamePasswordBind.USERNAMEKEY + "' >> auth")
```

That does not rely on the shell expansion or shell quoting and will still write the value of those two environment variables to the output file for later use in assertions.  I think it will significantly simplify the test and let us avoid all the complexity of using an `ant` step.